File: MatrixShader/src/MatrixShader.Core/Services/LicenseService.cs
using System.Net.Http;
using System.Text;

namespace MatrixShader.Core.Services;

/// <summary>
/// License service — server-authoritative validation.
/// Client checks if a key file exists on disk (offline use after activation).
/// Activation sends key to server for HMAC validation + machine limit check.
/// The client NEVER has the signing secret.
/// </summary>
public sealed class LicenseService : ILicenseService
{
    private static readonly string LicenseDir = Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
        "MatrixShader");

    private static readonly string LicensePath = Path.Combine(LicenseDir, "license.key");

    private const string ValidateUrl = "https://matrixshader.com/api/validate";
    private static readonly TimeSpan ServerTimeout = TimeSpan.FromSeconds(8);

    private bool? _cachedResult;

    /// <inheritdoc/>
    public bool IsLicensed
    {
        get
        {
            if (_cachedResult.HasValue)
                return _cachedResult.Value;

            // Licensed = key file exists with valid format on disk.
            // The key was validated by the server during activation.
            var key = GetInstalledKey();
            _cachedResult = key != null && HasValidFormat(key);
            return _cachedResult.Value;
        }
    }

    /// <inheritdoc/>
    public ActivationResult Activate(string key)
    {
        if (!HasValidFormat(key))
            return ActivationResult.InvalidKey;

        // Server does the real validation (HMAC + machine limit)
        var serverResult = CheckServerActivation(key);
        if (serverResult != ActivationResult.Success)
            return serverResult;

        try
        {
            Directory.CreateDirectory(LicenseDir);
            File.WriteAllText(LicensePath, key.Trim().ToUpperInvariant());
            _cachedResult = true;
            DiagnosticLogger.Info("LICENSE", "License activated successfully");
            return ActivationResult.Success;
        }
        catch (Exception ex)
        {
            DiagnosticLogger.Error("LICENSE", $"Failed to save license: {ex.Message}");
            return ActivationResult.SaveFailed;
        }
    }

    /// <inheritdoc/>
    public string? GetInstalledKey()
    {
        try
        {
            if (!File.Exists(LicensePath))
                return null;

            var key = File.ReadAllText(LicensePath).Trim();
            return string.IsNullOrWhiteSpace(key) ? null : key;
        }
        catch
        {
            return null;
        }
    }

    /// <summary>
    /// Checks format only — REDPILL-XXXX-XXXX-XXXX-XXXX.
    /// Does NOT validate the HMAC signature (server does that).
    /// </summary>
    public bool ValidateKey(string key) => HasValidFormat(key);

    private static bool HasValidFormat(string key)
    {
        if (string.IsNullOrWhiteSpace(key))
            return false;

        key = key.Trim().ToUpperInvariant();
        var parts = key.Split('-');
        if (parts.Length != 5)
            return false;

        if (parts[0] != "REDPILL")
            return false;

        for (int i = 1; i <= 4; i++)
        {
            if (parts[i].Length != 4)
                return false;
            if (!parts[i].All(c => char.IsLetterOrDigit(c)))
                return false;
        }

        return true;
    }

    /// <summary>
    /// Calls /api/validate to check key validity + activation count.
    /// The SERVER does HMAC validation — client never has the secret.
    /// </summary>
    private static ActivationResult CheckServerActivation(string key)
    {
        try
        {
            using var client = new HttpClient { Timeout = ServerTimeout };

            var fingerprint = MachineFingerprint.Get();
            var normalizedKey = key.Trim().ToUpperInvariant();
            var payload = $"{{\"key\":\"{normalizedKey}\",\"fingerprint\":\"{fingerprint}\"}}";

            var content = new StringContent(payload, Encoding.UTF8, "application/json");
            var response = client.PostAsync(ValidateUrl, content).GetAwaiter().GetResult();

            if (response.StatusCode == System.Net.HttpStatusCode.Forbidden)
            {
                DiagnosticLogger.Warn("LICENSE", "Activation limit exceeded for this key");
                return ActivationResult.ActivationLimitExceeded;
            }

            if (response.IsSuccessStatusCode)
            {
                DiagnosticLogger.Info("LICENSE", $"Server validation: {(int)response.StatusCode}");
                return ActivationResult.Success;
            }

            DiagnosticLogger.Warn("LICENSE", $"Server error {(int)response.StatusCode}, activation blocked");
            return ActivationResult.ServerUnreachable;
        }
        catch (Exception ex)
        {
            DiagnosticLogger.Warn("LICENSE", $"Server unreachable: {ex.Message}");
            return ActivationResult.ServerUnreachable;
        }
    }
}

File: linux/license_service.py
"""License validation with server-side activation tracking.

Key format: REDPILL-XXXX-XXXX-XXXX-XXXX

Design philosophy: honest people pay, pirates never would have.
Don't punish paying customers with aggressive DRM.
First activation requires server verification to enforce machine limits.
After activation, license is fully offline — no phone-home ever.

Security: HMAC validation is server-only. The client never sees the
signing secret. The activation server is the sole authority on whether
a key is valid.
"""

import os
import sys
from enum import Enum

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import machine_fingerprint


# ---------------------------------------------------------------------------
# License storage
# ---------------------------------------------------------------------------

LICENSE_DIR = os.path.join(
    os.environ.get("XDG_CONFIG_HOME", os.path.expanduser("~/.config")),
    "matrix-shader",
)
LICENSE_PATH = os.path.join(LICENSE_DIR, "license.key")

VALIDATE_URL = "https://matrixshader.com/api/validate"
SERVER_TIMEOUT = 8  # seconds


# ---------------------------------------------------------------------------
# ActivationResult enum — mirrors C# ActivationResult
# ---------------------------------------------------------------------------

class ActivationResult(Enum):
    SUCCESS = "success"
    INVALID_KEY = "invalid_key"
    ACTIVATION_LIMIT_EXCEEDED = "activation_limit_exceeded"
    SERVER_UNREACHABLE = "server_unreachable"
    SAVE_FAILED = "save_failed"


# ---------------------------------------------------------------------------
# Core functions
# ---------------------------------------------------------------------------

def _looks_like_key(key):
    """Check if a string has valid license key format (no HMAC check).

    Returns True if key format is REDPILL-XXXX-XXXX-XXXX-XXXX with
    alphanumeric groups of 4 characters each.
    """
    if not key or not key.strip():
        return False

    key = key.strip().upper()
    parts = key.split("-")

    if len(parts) != 5:
        return False
    if parts[0] != "REDPILL":
        return False

    for i in range(1, 5):
        if len(parts[i]) != 4:
            return False
        if not parts[i].isalnum():
            return False

    return True


def get_installed_key():
    """Get the installed license key, or None if none exists."""
    try:
        if not os.path.exists(LICENSE_PATH):
            return None
        with open(LICENSE_PATH) as f:
            key = f.read().strip()
        return key if key else None
    except OSError:
        return None


def is_licensed():
    """Check if a valid Red Pill license is installed.

    After server-verified activation, the key file on disk is proof
    of activation. No client-side HMAC check — the server already
    validated the key during activation.
    """
    key = get_installed_key()
    return key is not None and _looks_like_key(key)


def _check_server_activation(key):
    """Call /api/validate to register activation with the server.

    The server performs HMAC validation and tracks machine activations.
    Returns SUCCESS if server confirms, ACTIVATION_LIMIT_EXCEEDED if over limit,
    INVALID_KEY if HMAC fails, or SERVER_UNREACHABLE if the server cannot be contacted.
    """
    try:
        import urllib.request
        import urllib.error
        import json

        fingerprint = machine_fingerprint.get()
        normalized = key.strip().upper()
        payload = json.dumps({"key": normalized, "fingerprint": fingerprint})

        req = urllib.request.Request(
            VALIDATE_URL,
            data=payload.encode("utf-8"),
            headers={"Content-Type": "application/json"},
            method="POST",
        )

        try:
            urllib.request.urlopen(req, timeout=SERVER_TIMEOUT)
        except urllib.error.HTTPError as e:
            if e.code == 403:
                try:
                    body = json.loads(e.read().decode())
                    if body.get("error") == "activation_limit":
                        return ActivationResult.ACTIVATION_LIMIT_EXCEEDED
                except Exception:
                    pass
                return ActivationResult.INVALID_KEY
            # Server error (500, 503, etc.) — don't let activation bypass the check
            return ActivationResult.SERVER_UNREACHABLE
        return ActivationResult.SUCCESS

    except Exception:
        # Network error, timeout, DNS failure — require connectivity for activation
        return ActivationResult.SERVER_UNREACHABLE


def activate(key):
    """Activate a license key via server validation.

    Checks key format locally (fast reject), then sends to server for
    HMAC verification and activation tracking. Saves key on success.
    Returns ActivationResult.
    """
    if not _looks_like_key(key):
        return ActivationResult.INVALID_KEY

    # Server-side activation check (required — no offline bypass)
    server_result = _check_server_activation(key)
    if server_result != ActivationResult.SUCCESS:
        return server_result

    try:
        os.makedirs(LICENSE_DIR, exist_ok=True)
        with open(LICENSE_PATH, "w") as f:
            f.write(key.strip().upper())
        return ActivationResult.SUCCESS
    except OSError:
        return ActivationResult.SAVE_FAILED
